flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -123,42 +134,48 @@ public TaskDeploymentDescriptor createDeploymentDescriptor(
                             resultId,
                             partitionType,
                             queueToRequest,
-                            getConsumedPartitionShuffleDescriptors(partitions)));
+                            getConsumedPartitionShuffleDescriptors(
+                                    consumedIntermediateResult, consumedPartitionGroup)));
         }
 
         return inputGates;
     }
 
-    private ShuffleDescriptor[] getConsumedPartitionShuffleDescriptors(
-            List<IntermediateResultPartition> partitions) {
+    private SerializedValue<ShuffleDescriptor[]> getConsumedPartitionShuffleDescriptors(
+            IntermediateResult intermediateResult, ConsumedPartitionGroup consumedPartitionGroup)
+            throws IOException {
+        SerializedValue<ShuffleDescriptor[]> serializedShuffleDescriptors =
+                intermediateResult.getCachedShuffleDescriptors(consumedPartitionGroup);
+        if (serializedShuffleDescriptors == null) {
+            serializedShuffleDescriptors =
+                    computeConsumedPartitionShuffleDescriptors(consumedPartitionGroup);
+            intermediateResult.cacheShuffleDescriptors(
+                    consumedPartitionGroup, serializedShuffleDescriptors);
+        }
+        return serializedShuffleDescriptors;
+    }
 
-        ShuffleDescriptor[] shuffleDescriptors = new ShuffleDescriptor[partitions.size()];
+    private SerializedValue<ShuffleDescriptor[]> computeConsumedPartitionShuffleDescriptors(
+            ConsumedPartitionGroup consumedPartitionGroup) throws IOException {
+
+        ShuffleDescriptor[] shuffleDescriptors =
+                new ShuffleDescriptor[consumedPartitionGroup.size()];
         // Each edge is connected to a different result partition
-        for (int i = 0; i < partitions.size(); i++) {
-            shuffleDescriptors[i] =
+        int i = 0;
+        for (IntermediateResultPartitionID partitionId : consumedPartitionGroup) {
+            shuffleDescriptors[i++] =
                     getConsumedPartitionShuffleDescriptor(
```

Comment:
Let's add some documents for the `UNKNOWN` `ShuffleDescriptor` code path to explain why we can cache it and do not need to update it.

Reply:
Documents added.